### PR TITLE
[Merged by Bors] - feat(MvPowerSeries.LexOrder): lexicographic order of multivariate power series

### PR DESCRIPTION
**Pull request by AntoineChambert-Loir**
Define the order of multivariate power series with respect to the lexicographic order on monomials.

Prove that the ring of multivariate power series over a domain is a domain.

Extension to some rings of Hahn series

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/MvPowerSeries/LexOrder.lean [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+section LexOrder
+
+open Finsupp
+variable [LinearOrder σ] [WellFoundedGT σ]
+
+/-- The lex order on multivariate power series.  -/
+noncomputable def lexOrder (φ : MvPowerSeries σ R) : (WithTop (Lex (σ →₀ ℕ))) := by
+  classical
+  exact if h : φ = 0 then ⊤ else by
+    have ne : Set.Nonempty (toLex '' φ.support)  := by
```
Comment (jcommelin) [inline Mathlib/RingTheory/MvPowerSeries/LexOrder.lean:33]:
```suggestion
    have ne : Set.Nonempty (toLex '' φ.support) := by
```

File: Mathlib/RingTheory/MvPowerSeries/LexOrder.lean [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+      lexOrder φ = WithTop.some ((@wellFounded_lt (Lex (σ →₀ ℕ))
+        (instLTLex) (Lex.wellFoundedLT)).min (toLex '' φ.support) ne) := by
+  suffices ne : Set.Nonempty (toLex '' φ.support) by
+    use ne
+    unfold lexOrder
+    simp only [dif_neg hφ]
+  simp only [Set.image_nonempty, Function.support_nonempty_iff, ne_eq, hφ, not_false_eq_true]
+
+theorem eq_zero_iff_lexOrder_eq_top (φ : MvPowerSeries σ R) :
+    φ = 0 ↔ lexOrder φ = ⊤ := by
```
Comment (jcommelin) [inline Mathlib/RingTheory/MvPowerSeries/LexOrder.lean:51]:
```suggestion
@[simp]
theorem lexOrder_eq_top_iff_eq_zero (φ : MvPowerSeries σ R) :
    lexOrder φ = ⊤ ↔ φ = 0 := by
```

File: Mathlib/RingTheory/MvPowerSeries/LexOrder.lean [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+theorem coeff_eq_zero_of_lt_lexOrder {φ : MvPowerSeries σ R} {d : σ →₀ ℕ}
+    (h : toLex d < lexOrder φ) : coeff R d φ = 0 := by
+  by_cases hφ : φ = 0
+  · simp only [hφ, map_zero]
+  · rcases lexOrder_def_of_ne_zero hφ with ⟨ne, hφ'⟩
+    rw [hφ', WithTop.coe_lt_coe] at h
+    by_contra h'
+    exact WellFounded.not_lt_min _ (toLex '' φ.support) ne (Set.mem_image_equiv.mpr h') h
+
+theorem lexOrder_le_of_coeff_neq_zero {φ : MvPowerSeries σ R} {d : σ →₀ ℕ}
```
Comment (jcommelin) [inline Mathlib/RingTheory/MvPowerSeries/LexOrder.lean:89]:
```suggestion
theorem lexOrder_le_of_coeff_ne_zero {φ : MvPowerSeries σ R} {d : σ →₀ ℕ}
```

File: Mathlib/RingTheory/MvPowerSeries/NoZeroDivisors.lean [hunk 4]
```diff
@@ -4,8 +4,9 @@ Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Antoine Chambert-Loir
 -/
 
-import Mathlib.RingTheory.MvPowerSeries.Basic
+-- import Mathlib.RingTheory.MvPowerSeries.Basic
```
Comment (jcommelin) [inline Mathlib/RingTheory/MvPowerSeries/NoZeroDivisors.lean:7]:
```suggestion
```

File: Mathlib/RingTheory/MvPowerSeries/NoZeroDivisors.lean [hunk 5]
```diff
@@ -65,6 +66,23 @@ theorem mem_nonZeroDivisors_of_constantCoeff {φ : MvPowerSeries σ R}
...
+
+variable {σ R : Type*} [Semiring R] [NoZeroDivisors R]
+
+instance : NoZeroDivisors (MvPowerSeries σ R) where
+  eq_zero_or_eq_zero_of_mul_eq_zero {φ ψ} h := by
+    letI : LinearOrder σ := LinearOrder.swap σ WellOrderingRel.isWellOrder.linearOrder
+    letI : WellFoundedGT σ := by
+      unfold WellFoundedGT
+      suffices IsWellFounded σ fun x y ↦ WellOrderingRel x y by
+        exact this
```
Comment (jcommelin) [inline Mathlib/RingTheory/MvPowerSeries/NoZeroDivisors.lean:79]: Do you need these lines? Or is it enough to have
```suggestion
      change IsWellFounded σ fun x y ↦ WellOrderingRel x y
```

Comment (AntoineChambert-Loir) [reply Mathlib/RingTheory/MvPowerSeries/NoZeroDivisors.lean:78]: K

Comment (AntoineChambert-Loir) [reply Mathlib/RingTheory/MvPowerSeries/NoZeroDivisors.lean:78]: OK.

**Review by jcommelin — COMMENTED** [review]
I think this is almost ready to go.

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

File: Mathlib/RingTheory/HahnSeries/PowerSeries.lean [hunk 6]
```diff
@@ -169,6 +175,11 @@ def toMvPowerSeries {σ : Type*} [Finite σ] : HahnSeries (σ →₀ ℕ) R ≃+
 
 variable {σ : Type*} [Finite σ]
 
+-- TODO : generalize to all (?) rings of Hahn Series
+/-- If R has no zero divisors, then `HahnSeries (σ →₀ ℕ) R` has no zero divisors, for finite `σ` -/
```
Comment (bryangingechen) [inline Mathlib/RingTheory/HahnSeries/PowerSeries.lean:179]: Strange. [bors CI seems to think this line is over the 100 character limit](https://github.com/leanprover-community/mathlib4/actions/runs/10905668118/job/30265082041#step:11:2484), but it seemed to be fine here.

**Comment by bryangingechen** [comment]
Could you try merging `master`, fixing the build and then putting this back on the queue? Thanks!
bors r-
bors d+

**Comment by bryangingechen** [comment]
Thanks, looks like CI should be happy with this now!
bors r+

**Pull request closed without merging**